tests/test_feedexport.py

```diff
@@ -1587,16 +1590,60 @@ def test_export_items(self):
         header = self.MyItem.fields.keys()
         yield self.assertExported(items, header, rows, settings=Settings(settings))
 
+    @defer.inlineCallbacks
+    def test_export_file_size_trigger(self):
+        items = [
+            self.MyItem({'foo': 'bar1', 'egg': 'spam1', 'baz': 'quux1'}),
+            self.MyItem({'foo': 'bar2', 'egg': 'spam2', 'baz': 'quux2'}),
+        ]
+        fmt_to_limits = {
+            'json': ('40B', '90B'),
+            'jsonlines': ('40B', '90B'),
+            'xml': ('100B', '160B'),
+            'csv': ('30B', '40B'),
+            'marshal': ('30B', '70B'),
+            'pickle': ('50B', '110B'),
+        }
+        for fmt, limits in fmt_to_limits.items():
+            #  +1 for empty files after batch has been triggered
+            for limit, expected in {'0B': 1, '1B': 2 + 1, limits[0]: 2 + 1, limits[1]: 1 + 1}.items():
+                settings = {
+                    'FEEDS': {
+                        os.path.join(self._random_temp_filename(), fmt, self._file_mark): {'format': fmt},
+                    },
+                    'FEED_EXPORT_BATCH_FILE_SIZE': limit,
+                }
+                data = yield self.exported_data(items, settings)
+                self.assertEqual(expected, len(data[fmt]))
+
+    @defer.inlineCallbacks
+    def test_export_duration_trigger(self):
+        items = [self.MyItem({'foo': 'bar1', 'egg': 'spam1', 'baz': 'quux1'})]
+        for fmt in ('csv', 'json', 'jsonlines', 'marshal', 'pickle'):
+            #  +1 for empty files after batch has been triggered
+            for delay, expected in {'0:0:0': 1, '0:0:0.01': 2 + 1, '0:0:0.1': 2 + 1, '0:0:1.2': 1}.items():
```

Comment:
Why is the last value 1 instead of 1+1? Why is the behavior with duration different than with file size above?

Reply:
Ok yeah it should be trigger after getting at most 2 items. I will change that.

Reply:
![image](https://user-images.githubusercontent.com/23308719/128304487-ac8f08f0-5b09-4903-8bca-8e5b909dbe1f.png)

In the last entry, I was expecting item being added after a delay of 0.5 seconds at least but it was added sooner. So the batch wasn't triggered and there was only 1 batch instead of 1 (with 2 items) + 1 empty batch. Kinda hard to get at-most-2-items batch everytime. Should I continue with expecting 1 batch with 2 items duration?

Reply:
Maybe you need to also limit concurrency to 1? It’s 8 by default, so multiple requests can happen in parallel.

Reply:
Tried `CONCURRENT_REQUESTS: 1` still getting some results with delays less than 0.5 secs :(

Reply:
Can you share code that I can use to reproduce the issue easily (i.e. what you used to print the text in the screenshot) in some branch of your Scrapy fork?

Reply:
[https://github.com/drs-11/scrapy/tree/batch-duration-debug](https://github.com/drs-11/scrapy/tree/batch-duration-debug)

I did the usual one item per url and two urls. The wrong delays does not happen everytime so you may need to run the crawl multiple times.

Reply:
Oh, I bet it’s https://docs.scrapy.org/en/latest/topics/settings.html#std-setting-RANDOMIZE_DOWNLOAD_DELAY

Reply:
YES! Getting consistent delays now :sweat_smile:.